internal/replication/env_test.go

```diff
@@ -0,0 +1,396 @@
+
+	for n := uint32(1); n <= uint32(nNodes); n++ {
+		nEnv, err := newNodeEnv(n, testDir, lg, clusterConfig)
+		if err != nil {
+			os.RemoveAll(testDir)
+			return nil
+		}
+
+		cEnv.nodes = append(cEnv.nodes, nEnv)
+	}
+
+	return cEnv
+}
+
+func newNodeEnv(n uint32, testDir string, lg *logger.SugarLogger, clusterConfig *types.ClusterConfig) (*nodeEnv, error) {
+
+	nodeID := fmt.Sprintf("node%d", n)
+	localTestDir := path.Join(testDir, nodeID)
+
+	localConf := &config.LocalConfiguration{
+		Server: config.ServerConf{
+			Identity: config.IdentityConf{
+				ID: nodeID,
+			},
+		},
+		Replication: config.ReplicationConf{
+			WALDir:  path.Join(testDir, nodeID, "wal"),
+			SnapDir: path.Join(testDir, nodeID, "snap"),
+			Network: config.NetworkConf{
+				Address: "127.0.0.1",
+				Port:    peerPortBase + n,
+			},
+			TLS: config.TLSConf{
+				Enabled: false,
+			},
+		},
+	}
+
+	peerTransport := comm.NewHTTPTransport(&comm.Config{
+		LocalConf: localConf,
+		Logger:    lg,
+	})
+
+	qBarrier := queue.NewOneQueueBarrier(lg)
+
+	ledger := &memLedger{}
+	proposedBlock := &types.Block{
+		Header: &types.BlockHeader{
+			BaseHeader: &types.BlockHeaderBase{
+				Number: 1,
+			},
+		},
+	}
+	if err := ledger.Append(proposedBlock); err != nil { //genesis block
+		return nil, err
+	}
+
+	conf := &replication.Config{
+		LocalConf:            localConf,
+		ClusterConfig:        clusterConfig,
+		LedgerReader:         ledger,
+		Transport:            peerTransport,
+		BlockOneQueueBarrier: qBarrier,
+		Logger:               lg,
+	}
+
+	blockReplicator, err := replication.NewBlockReplicator(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	err = conf.Transport.SetConsensusListener(blockReplicator)
+	if err != nil {
+		return nil, err
+	}
+
+	err = conf.Transport.UpdateClusterConfig(conf.ClusterConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	env := &nodeEnv{
+		testDir:         localTestDir,
+		conf:            conf,
+		blockReplicator: blockReplicator,
+		ledger:          ledger,
+		stopServeCh:     make(chan struct{}),
+	}
+
+	return env, nil
+}
+
+// find the index [0,N) of the leader node, -1 if no leader.
+func (c *clusterEnv) FindLeaderIndex() int {
+	for idx, e := range c.nodes {
+		leader := e.blockReplicator.IsLeader()
+		if leader == nil {
+			return idx
+		}
+	}
+
+	return -1
+}
+
+// find the index [0,N) of the leader node, if all indices agree; -1 if no agreed leader.
+func (c *clusterEnv) AgreedLeaderIndex(indices ...int) int {
+	if len(indices) == 0 {
+		for i := 0; i < len(c.nodes); i++ {
+			indices = append(indices, i)
+		}
+	}
+	leaderIdx := c.FindLeaderIndex()
+	if leaderIdx < 0 {
+		return leaderIdx
+	}
+
+	leaderRaftID := c.nodes[leaderIdx].blockReplicator.RaftID()
+
+	for _, idx := range indices {
+		node := c.nodes[idx]
+		leaderErr := node.blockReplicator.IsLeader()
+		if leaderErr == nil {
+			if leaderIdx != idx {
+				return -1
+			}
+		} else {
+			if leaderErr.GetLeaderID() != leaderRaftID {
+				return -1
+			}
+		}
+	}
```

Comment:
`IsLedger()` is creating confusion here. 
```suggestion
	for _, idx := range indices {
		node := c.nodes[idx]
		if node.blockReplicator.GetLeader() != leaderRaftID {
			return -1
		}
	}
```

Reply:
Good suggestion.